docs/branching-model.md

```diff
@@ -0,0 +1,57 @@
+# Git branching model
+
+## Master branch
+
+The default branch for a project is the master branch. This branch must always be in a stable state: the build must be successful, all tests must pass, style must follow the [style guide](cplusplus-style-guide.md) and static analysis should not give warnings or errors.
+
+Commits representing a release will be tagged as described in the [release strategy](release-strategy.md#tagging-and-version-numbers).
+
+
+## Development forks
+
+In order to keep the main repo clear from development branches, developers must create a fork of any project they are planning on contributing to. In addition, for projects which are or will be open sourced this is the easiest way to begin developing.
+The only scenario when it may be currently necessary to push branches directly to the repository is where the branch changes the Jenkinsfile. These should not be picked up from a fork for security reasons.
+
+Once the fork has been created, it is expected that the user would create a branch to track each new feature, bugfix, etc,  being made. If any of these tasks are complex, they can be addressed in multiple increments that are individually reviewed and merged. See [Integration (pull requests)](#integration-pull-requests) for more information about merging development branches.
+
+A task is only considered done once all the branches associated with it have been merged into the master branch; this encourages tasks to be well defined so they can be assessed and merged quickly, with the minimum amount of code to be integrated every time.
+
+Long-lived development branches are strongly discouraged but if necessary, its creation must be discussed to try to find alternatives. Long-lived branches tend to strongly diverge from the master branch, making the review and integration complicated and tedious.
+
+To further aid clarity we advise using the following prefixes based on the nature of the change being made:
+
+ * `feature`: New feature
```

Comment:
I think these prefixes are too long for the information content they convey 7 characters here are carrying two bits of information. We should use abbreviations. "dev" for features "bug" for bugs. "backport" is rarer and could stay.

I strongly disagree with people using the word "refactor" at all because it's become an opaque term that spans everything between spellcorrecting comments through to large code changes. If it's a big change, it should be a "dev" branch, because it's altering code. If it's a small change, the prefix should describe the work carried out: "style" (code style changes), "doc" (documentation or commenting) or "move" (anything that involves moving code between files). Anything else is dev work and should have a goal, not a refactor - "refactor X" is not a task because it doesn't have a goal.

Reply:
> We should use abbreviations. "dev" for features "bug" for bugs. "backport" is rarer and could stay.

Many developers are familiar with `feature`, `bugfix`, etc because those are defined in widely used Git workflows. Having to type 2-3 extra characters once per day is not tragic and I have never been in the situation where it broke the Jira integration. 

The suggestion about splitting `refactor` in more specific categories sounds good a priori, but it doesn't always make sense to make each of those changes separately. I see 2 possible arguments:

1. In favor of splitting in 2 PRs: Documentation/style changes are much more straightforward to merge, so they could get in really quickly while moving move receives complaints :smile:
2. In favor of keeping them in 1 PR: If moving code modifies the structure of a class, and the documentation is adapted accordingly, then it would be more complicated to merge the changes in 2 steps.